packages/contracts-bedrock/scripts/autogen/generate-semver-lock/main.go

```diff
@@ -0,0 +1,141 @@
+package main
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+func main() {
+	if err := run(); err != nil {
+		panic(err)
+	}
+}
+
+func run() error {
+	// find semver files
+	files, err := findSemverFiles()
+	if err != nil {
+		return err
+	}
+
+	// Call the function to write semver lock (implementation needed)
+	if err := writeSemverLock(files); err != nil {
+		return fmt.Errorf("failed to write semver lock: %w", err)
+	}
+
+	return nil
+}
+
+func findSemverFiles() ([]string, error) {
+	// Execute grep command to find files with @custom:semver
+	var cmd = exec.Command("bash", "-c", "grep -rl '@custom:semver' src | jq -Rs 'split(\"\n\") | map(select(length > 0))'")
+	output, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse the JSON array of files
+	var files []string
+	if err := json.Unmarshal(output, &files); err != nil {
+		return nil, fmt.Errorf("failed to parse JSON output: %w", err)
+	}
+
+	return files, nil
+}
+
+func writeSemverLock(files []string) error {
+	// Map to store our JSON output
+	output := make(map[string]map[string]string)
+
+	for _, file := range files {
+		// Read file contents
+		fileContents, err := os.ReadFile(file)
+		if err != nil {
+			return fmt.Errorf("failed to read file %s: %w", file, err)
+		}
+
+		// Extract contract name from file path using regex
+		re := regexp.MustCompile(`src/.*/(.+)\.sol`)
```

Comment:
Can probably make things slightly cleaner by pulling variables like this out to above the `for` loop. Same with the code that gets the artifacts directory/config.